include/cuda/memory_resource

```diff
@@ -0,0 +1,519 @@
+    {
+      return {&_Resource_vtable_builder::_Alloc<_Resource>,
+              &_Resource_vtable_builder::_Dealloc<_Resource>,
+              &_Resource_vtable_builder::_Equal<_Resource>,
+              &_Resource_vtable_builder::_Alloc_async<_Resource>,
+              &_Resource_vtable_builder::_Dealloc_async<_Resource>};
+    }
+};
+// clang-format on
+
+template <_AllocType _Alloc_type, class _Resource>
+__device__ _LIBCUDACXX_INLINE_VAR constexpr _Async_alloc_vtable
+  __alloc_vtable = _Resource_vtable_builder::template _Create<_Resource, _Alloc_type>();
+
+template <class _Property>
+struct _Property_vtable
+{
+  using _PropertyFn         = typename _Property::value_type (*)(void*);
+  _PropertyFn __property_fn = nullptr;
+
+  _Property_vtable() = default;
+
+  _LIBCUDACXX_INLINE_VISIBILITY
+  constexpr _Property_vtable(_PropertyFn __property_fn_) noexcept
+      : __property_fn(__property_fn_)
+  {}
+};
+
+template <class... _Properties>
+struct _Resource_vtable : public _Property_vtable<_Properties>...
+{
+  _Resource_vtable() = default;
+
+  template <class... _PropertyFns>
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr _Resource_vtable(_PropertyFns... __property_fn_) noexcept
+      : _Property_vtable<_Properties>(__property_fn_)...
+  {}
+
+  template <class _Resource>
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr _Resource_vtable _Create() noexcept
+  {
+    return {&_Resource_vtable_builder::_Get_property<_Resource, _Properties>...};
+  }
+};
+
+template <class... _Properties>
+struct _Filtered;
+
+template <bool _IsUniqueProperty>
+struct _Property_filter
+{
+  template <class _Property, class... _Properties>
+  using _Filtered_properties =
+    typename _Filtered<_Properties...>::_Filtered_vtable::template _Append_property<_Property>;
+};
+template <>
+struct _Property_filter<false>
+{
+  template <class _Property, class... _Properties>
+  using _Filtered_properties = typename _Filtered<_Properties...>::_Filtered_vtable;
+};
+
+template <class _Property, class... _Properties>
+struct _Filtered<_Property, _Properties...>
+{
+  using _Filtered_vtable =
+    typename _Property_filter<property_with_value<_Property> && !_CUDA_VSTD::_One_of<_Property, _Properties...>>::
+      template _Filtered_properties<_Property, _Properties...>;
+
+  template <class _OtherPropery>
+  using _Append_property = _Filtered<_OtherPropery, _Property, _Properties...>;
+
+  using _Vtable = _Resource_vtable<_Property, _Properties...>;
+};
+
+template <>
+struct _Filtered<>
+{
+  using _Filtered_vtable = _Filtered<>;
+
+  template <class _OtherPropery>
+  using _Append_property = _Filtered<_OtherPropery>;
+
+  using _Vtable = _Resource_vtable<>;
+};
+
+template <class... _Properties>
+using _Filtered_vtable = typename _Filtered<_Properties...>::_Filtered_vtable::_Vtable;
+
+template <_AllocType _Alloc_type, class... _Properties> //
+class basic_resource_ref : protected _Filtered_vtable<_Properties...>
```

Comment:
I'd suggest moving the allocation interface to another base class. This will reduce compilation time, object file size and debug binary size - also, I think that having the base members (object + vtable) at the beginning may have additional benefits, especially when interfacing with C.

Reply:
Done